spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/inbound/AbstractHazelcastMessageProducer.java

```diff
@@ -0,0 +1,198 @@
+
+	protected AbstractHazelcastMessageProducer(DistributedObject distributedObject) {
+		Assert.notNull(distributedObject, "cache must not be null");
+		this.distributedObject = distributedObject;
+	}
+
+	@Override
+	protected void onInit() {
+		super.onInit();
+		HazelcastIntegrationDefinitionValidator.validateCacheEventsByDistributedObject(
+				getDistributedObject(), getCacheEventTypeSet());
+	}
+
+	protected Set<String> getCacheEventTypeSet() {
+		return StringUtils.commaDelimitedListToSet(getCacheEventTypes());
+	}
+
+	protected DistributedObject getDistributedObject() {
+		return distributedObject;
+	}
+
+	protected String getCacheEventTypes() {
+		return cacheEventTypes;
+	}
+
+	public void setCacheEventTypes(String cacheEventTypes) {
+		HazelcastIntegrationDefinitionValidator.validateEnumType(CacheEventType.class, cacheEventTypes);
+		this.cacheEventTypes = cacheEventTypes;
+	}
+
+	protected CacheListeningPolicyType getCacheListeningPolicy() {
+		return cacheListeningPolicy;
+	}
+
+	public void setCacheListeningPolicy(CacheListeningPolicyType cacheListeningPolicy) {
+		this.cacheListeningPolicy = cacheListeningPolicy;
+	}
+
+	protected String getHazelcastRegisteredEventListenerId() {
+		return hazelcastRegisteredEventListenerId;
+	}
+
+	protected void setHazelcastRegisteredEventListenerId(String hazelcastRegisteredEventListenerId) {
+		this.hazelcastRegisteredEventListenerId = hazelcastRegisteredEventListenerId;
+	}
+
+	protected abstract class AbstractHazelcastEventListener {
+
+		protected abstract void processEvent(EventObject event);
+
+		protected void sendMessage(EventObject event, InetSocketAddress socketAddress,
+				CacheListeningPolicyType cacheListeningPolicyType) {
+			if (CacheListeningPolicyType.ALL == cacheListeningPolicyType || isEventAcceptable(socketAddress)) {
+				AbstractHazelcastMessageProducer.this.sendMessage(getMessageBuilderFactory().withPayload(event).build());
+			}
+		}
+
+		private boolean isEventAcceptable(InetSocketAddress socketAddress) {
+			final Set<SocketAddress> localSocketAddressesSet = getLocalSocketAddresses();
+			if ((!localSocketAddressesSet.isEmpty())
+					&& (localSocketAddressesSet.contains(socketAddress) ||
+							isEventComingFromNonRegisteredHazelcastInstance(localSocketAddressesSet, socketAddress))) {
+				return true;
+			}
+
+			return false;
+		}
+
+		private Set<SocketAddress> getLocalSocketAddresses() {
+			final Set<SocketAddress> localSocketAddressesSet = new HashSet<>();
+			for (HazelcastInstance hazelcastInstance : Hazelcast.getAllHazelcastInstances()) {
+				localSocketAddressesSet.add(hazelcastInstance.getLocalEndpoint().getSocketAddress());
+			}
+
+			return localSocketAddressesSet;
+		}
+
+		private boolean isEventComingFromNonRegisteredHazelcastInstance(
+				Set<SocketAddress> localSocketAddressesSet, InetSocketAddress socketAddressOfEvent) {
+			final HazelcastInstance hazelcastInstance = Hazelcast.getAllHazelcastInstances().iterator().next();
```

Comment:
Can't we reuse `Hazelcast.getAllHazelcastInstances()` from that `getLocalSocketAddresses()` for better performance?